storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java

```diff
@@ -0,0 +1,1146 @@
+                            config,
+                            time,
+                            false,
+                            config.initFileSize(),
+                            config.preallocate,
+                            "");
+                    segments.add(newSegment);
+
+                    // We need to update the segment base offset and append position data of the metadata when log rolls.
+                    // The next offset should not change.
+                    updateLogEndOffset(nextOffsetMetadata.messageOffset);
+
+                    logger.info("Rolled new log segment at offset {} in {} ms.", newOffset, time.hiResClockMs() - start);
+
+                    return newSegment;
+                });
+    }
+
+
+    /**
+     * Delete all data in the local log and start at the new offset.
+     *
+     * @param newOffset The new offset to start the log with
+     * @return the list of segments that were scheduled for deletion
+     */
+    public Iterable<LogSegment> truncateFullyAndStartAt(long newOffset) {
+        return maybeHandleIOException("Error while truncating the entire log for " + topicPartition + " in dir " + dir.getParent(),
+                () -> {
+                    logger.debug("Truncate and start at offset {}", newOffset);
+                    checkIfMemoryMappedBufferClosed();
+                    List<LogSegment> segmentsToDelete = new ArrayList<>(segments.values());
+
+                    if (!segmentsToDelete.isEmpty()) {
+                        removeAndDeleteSegments(segmentsToDelete.subList(0, segmentsToDelete.size() - 1), true, new LogTruncation(this.logger));
+                        // Use createAndDeleteSegment() to create new segment first and then delete the old last segment to prevent missing
+                        // active segment during the deletion process
+                        createAndDeleteSegment(newOffset, segmentsToDelete.get(segmentsToDelete.size() - 1), true, new LogTruncation(this.logger));
+                    }
+
+                    updateLogEndOffset(newOffset);
+
+                    return segmentsToDelete;
+                });
+    }
+
+    /**
+     * Truncate this log so that it ends with the greatest offset < targetOffset.
+     *
+     * @param targetOffset The offset to truncate to, an upper bound on all offsets in the log after truncation is complete.
+     * @return the list of segments that were scheduled for deletion
+     */
+    public Iterable<LogSegment> truncateTo(long targetOffset) throws IOException {
+        Collection<LogSegment> deletableSegments = segments.filter(segment -> segment.baseOffset() > targetOffset);
+        removeAndDeleteSegments(deletableSegments, true, new LogTruncation(this.logger));
+        segments.activeSegment().truncateTo(targetOffset);
+        updateLogEndOffset(targetOffset);
+        return deletableSegments;
+    }
+
+    private <T> T maybeHandleIOException(String errorMsg,
```

Comment:
`errorMsg` should be a `Supplier<String>` so we only do string concat when needed. The Scala code did that since it used a by-name argument.